packages/app/src-tauri/src/commands/ftp.rs

```diff
@@ -1,105 +1,216 @@
+use uuid::Uuid;
+
+use crate::{
+    log_if_error_and_return,
+    utils::{
+        event::{Channel, Event, EventProducer},
+        file::{get_file_info, write_string_to_temp_file, TempFile},
+        result::Result,
+    },
+};
+
+use super::common::{EndPointPayloadConf, FileInfo};
+
+#[derive(Deserialize)]
+struct FtpConnection {
     host: String,
     port: u16,
     user: String,
     password: String,
+}
+
+impl FtpConnection {
+    async fn connect(&self) -> Result<AsyncFtpStream> {
+        let addr = format!("{}:{}", self.host, self.port);
+        let mut ftp_stream = AsyncFtpStream::connect(addr).await?;
+        ftp_stream.login(&self.user, &self.password).await?;
+
+        // As default set the FTP connection to passive mode
+        ftp_stream.set_mode(Mode::Passive);
+        // https://www.iana.org/assignments/ftp-commands-extensions/ftp-commands-extensions.xhtml
+        if ftp_stream
+            .feat()
+            .await
+            .is_ok_and(|opts| opts.get("EPSV").is_some())
+        {
+            ftp_stream.set_mode(Mode::ExtendedPassive);
+        }
+
+        Ok(ftp_stream)
+    }
+}
+
+#[derive(Deserialize)]
+pub struct FilePayload {
     local_path: String,
-    path: Option<String>,
-    file_name: String,
-    http_host: String,
-    https: bool,
+
+    #[serde(flatten)]
+    connection: FtpConnection,
+
+    #[serde(flatten)]
+    endpoint_config: EndPointPayloadConf,
 }
 
-#[derive(serde::Deserialize)]
+#[derive(Deserialize)]
 pub struct XmlPayload {
-    host: String,
-    port: u16,
-    user: String,
-    password: String,
     content: String,
-    path: Option<String>,
-    file_name: String,
-    http_host: String,
-    https: bool,
-}
 
-#[derive(serde::Serialize)]
-pub struct FileInfo {
-    pub url: String,
-    pub mime_type: String,
-    pub size: u64,
+    #[serde(flatten)]
+    connection: FtpConnection,
+
+    #[serde(flatten)]
+    endpoint_config: EndPointPayloadConf,
 }
 
 #[named]
 #[tauri::command]
-pub async fn ftp_upload(payload: FilePayload) -> Result<FileInfo> {
-    let upload_result = ftp_upload_internal(payload).await;
-    log_if_error_and_return!(upload_result)
+pub async fn ftp_xml_upload_window_progress(window: Window, payload: XmlPayload) -> Result<Uuid> {
+    let result = ftp_xml_upload_with_progress_internal(window.into(), payload).await;
+    log_if_error_and_return!(result)
 }
 
-#[named]
 #[tauri::command]
-pub async fn ftp_xml_upload(payload: XmlPayload) -> Result<FileInfo> {
-    let app_data_path = data_dir().unwrap(); //improve error handling
-    let p = format!("{}{}", app_data_path.to_string_lossy(), "/feed.xml");
+pub async fn ftp_upload_window_progress(window: Window, payload: FilePayload) -> Uuid {
+    ftp_upload_progress_internal(window.into(), payload, None)
+}
 
-    let xml = payload.content.as_bytes();
-    fs::write(p.clone(), xml).await?;
+async fn ftp_xml_upload_with_progress_internal(
+    channel: Channel,
```

Comment:
I don't understand why in line 76 we are passing `window` of type window and here we get `channel` of type `Channel`